src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,314 @@
+  }
+
+  private Message maybeAddToCache(Message message) {
+    Optional.ofNullable(caches.get(message.getQuestion().getDClass()))
+        .ifPresent(cache -> cache.addMessage(message));
+    return message;
+  }
+
+  private CompletionStage<LookupResult> setResponseToMessageFuture(SetResponse setResponse) {
+    if (setResponse.isNXDOMAIN()) {
+      return completeExceptionally(new NoSuchDomainException());
+    }
+    if (setResponse.isNXRRSET()) {
+      return completeExceptionally(new NoSuchRRSetException());
+    }
+    if (setResponse.isSuccessful()) {
+      List<Record> records =
+          setResponse.answers().stream()
+              .flatMap(rrset -> rrset.rrs(cycleResults).stream())
+              .collect(toList());
+      return completedFuture(new LookupResult(records, null));
+    }
+    return null;
+  }
+
+  private <T extends LookupFailedException> CompletionStage<LookupResult> completeExceptionally(
+      T failure) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    future.completeExceptionally(failure);
+    return future;
+  }
+
+  private CompletionStage<LookupResult> resolveRedirects(LookupResult response, Record query) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    maybeFollowRedirect(response, query, 1, future);
+    return future;
+  }
+
+  private void maybeFollowRedirect(
+      LookupResult response,
+      Record query,
+      int redirectCount,
+      CompletableFuture<LookupResult> future) {
+    try {
+      if (redirectCount > maxRedirects) {
+        throw new RedirectOverflowException(
+            format("Refusing to follow more than %s redirects", maxRedirects));
+      }
+
+      List<Record> records = response.getRecords();
+      if (records.isEmpty()) {
+        future.complete(response);
+      } else if (records.get(0).getType() == DNAME || records.get(0).getType() == CNAME) {
+        lookupWithCache(
+                buildRedirectQuery(response, query),
+                makeAliases(response.getAliases(), query.getName()))
+            .thenAccept(m -> maybeFollowRedirect(m, query, redirectCount + 1, future));
+      } else {
+        future.complete(response);
+      }
+    } catch (LookupFailedException e) {
+      future.completeExceptionally(e);
+    }
+  }
+
+  /** Return an unmodifiable list containing the contents of previous, if any, plus name. */
+  private List<Name> makeAliases(List<Name> previous, Name name) {
+    if (previous == null) {
+      return singletonList(name);
+    } else {
+      List<Name> copy = new ArrayList<>(previous);
+      copy.add(name);
+      return Collections.unmodifiableList(copy);
+    }
+  }
+
+  private Record buildRedirectQuery(LookupResult response, Record question) {
+    List<Record> answer = response.getRecords();
+    Record firstAnswer = answer.get(0);
+    if (answer.size() != 1) {
+      throw new InvalidZoneDataException("Multiple CNAME RRs not allowed, SEE RFC1034 3.6.2");
```

Comment:
`see` or is there a reason for writing this uppercase?

Reply:
No reason 🤷